packages/dds/merge-tree/src/mergeTree.ts

```diff
@@ -847,104 +836,154 @@ export class MergeTree {
-				}
+			for (const seg of segmentsWithSlidDst) {
+				cache?.set(seg, slideToSegment);
 			}
 			return slideToSegment;
 		}
+
 		// If no such segment is found, slide to the last valid segment.
 		backwardExcursion(segment, goFurtherToFindSlideToSegment);
-		if (shouldCache) {
-			for (const seg of segmentsWithSlidDst) {
-				this.cachedSlideDestination?.segmentToSlideDestination.set(
-					seg,
-					slideToSegment ?? "detached",
-				);
-			}
+		for (const seg of segmentsWithSlidDst) {
+			cache?.set(seg, slideToSegment ?? "detached");
 		}
+
 		return slideToSegment;
 	}
 
 	/**
-	 * This method should only be called when the current client sequence number is
+	 * Slides or removes references from the provided list of segments.
+	 * The order of the references is preserved.
+	 * @remarks -
+	 * 1. Preserving the order of the references is a useful property for reference-based undo/redo
+	 * (see revertibles.ts).
+	 * 2. For use cases which necessitate eventual consistency across clients,
+	 * this method should only be called with segments for which the current client sequence number is
 	 * max(remove segment sequence number, add reference sequence number).
-	 * Otherwise eventual consistency is not guaranteed.
 	 * See `packages\dds\merge-tree\REFERENCEPOSITIONS.md`
+	 * @param segments - An array of (not necessarily contiguous) segments with increasing ordinals.
 	 */
-	private slideAckedRemovedSegmentReferences(segment: ISegment) {
-		assert(
-			isRemovedAndAcked(segment),
-			0x2f1 /* slideReferences from a segment which has not been removed and acked */,
-		);
-		if (segment.localRefs?.empty !== false) {
-			return;
-		}
-		const newSegment = this._getSlideToSegment(segment, true);
-		if (newSegment) {
-			const localRefs = (newSegment.localRefs ??= new LocalReferenceCollection(newSegment));
-			if (newSegment.ordinal < segment.ordinal) {
-				localRefs.addAfterTombstones(segment.localRefs);
-			} else {
-				localRefs.addBeforeTombstones(segment.localRefs);
-			}
-		} else {
-			for (const ref of segment.localRefs) {
+	private slideAckedRemovedSegmentReferences(segments: ISegment[]) {
+		const detachRefs = (segment: ISegment) => {
+			for (const ref of segment.localRefs ?? []) {
 				if (!refTypeIncludesFlag(ref, ReferenceType.StayOnRemove)) {
 					ref.callbacks?.beforeSlide?.(ref);
 					segment.localRefs?.removeLocalRef(ref);
 					ref.callbacks?.afterSlide?.(ref);
 				}
 			}
-		}
-		// TODO:AB#4069: This update might be avoidable by checking if the old segment
-		// had hierarchical refs before sliding using `segment.localRefs?.hierRefCount`.
-		if (newSegment) {
-			this.blockUpdatePathLengths(
-				newSegment.parent,
-				TreeMaintenanceSequenceNumber,
-				LocalClientId,
+		};
+		// References are slid in groups to preserve their order.
+		let currentSlideDestination: ISegment | undefined;
+		let currentSlideDirection: "forward" | "backward" | undefined;
+		let currentSlideGroup: ISegment[] = [];
```

Comment:
this can be simplified. rather than tracking segments, track the local reference collections: `segement.localRefs`. both add tombstone methods support tacking `...refs: Iterable<LocalReferencePosition>[]` so then you can just pass the whole array of tracked collections and the right things will happen, e.g. `addBeforeTombstones(...currentSlideGroup)`. once this is done you don't need to manually reverse the arrays.

Reply:
i changed it for my branch:
```
	private slideAckedRemovedSegmentReferences(segments: ISegment[]) {
		// References are slid in groups to preserve their order.
		let currentSlideDestination: ISegment | undefined;
		let currentSlideForward: boolean | undefined;
		let currentSlideGroup: LocalReferenceCollection[] = [];

		const slideGroup = () => {
			if (currentSlideDestination !== undefined) {
				const localRefs = (currentSlideDestination.localRefs ??=
					new LocalReferenceCollection(currentSlideDestination));
				if (currentSlideForward) {
					localRefs.addBeforeTombstones(...currentSlideGroup);
				} else {
					localRefs.addAfterTombstones(...currentSlideGroup);
				}

				// TODO:AB#4069: This update might be avoidable by checking if the old segment
				// had hierarchical refs before sliding using `segment.localRefs?.hierRefCount`.
				if (currentSlideDestination) {
					this.blockUpdatePathLengths(
						currentSlideDestination.parent,
						TreeMaintenanceSequenceNumber,
						LocalClientId,
					);
				}
			} else {
				for (const collection of currentSlideGroup) {
					for (const ref of collection) {
						if (!refTypeIncludesFlag(ref, ReferenceType.StayOnRemove)) {
							ref.callbacks?.beforeSlide?.(ref);
							collection.removeLocalRef(ref);
							ref.callbacks?.afterSlide?.(ref);
						}
					}
				}
			}
		};
		const segmentCache = new Map<ISegment, { seg?: ISegment }>();
		for (const segment of segments) {
			assert(
				isRemovedAndAcked(segment),
				0x2f1 /* slideReferences from a segment which has not been removed and acked */,
			);
			if (segment.localRefs === undefined || segment.localRefs.empty) {
				continue;
			}
			const slideToSegment = this._getSlideToSegment(segment, segmentCache);
			const slideDirection =
				slideToSegment === undefined ? undefined : slideToSegment.ordinal > segment.ordinal;

			if (
				slideToSegment !== currentSlideDestination ||
				slideDirection !== currentSlideForward
			) {
				slideGroup();
				currentSlideGroup = [segment.localRefs];
				currentSlideDestination = slideToSegment;
				currentSlideForward = slideDirection;
			} else {
				currentSlideGroup.push(segment.localRefs);
			}
		}
		slideGroup();
	}

```